framework_laptop.c

```diff
@@ -335,15 +687,52 @@ static int framework_probe(struct platform_device *pdev)
 	}
 #endif
 
+	struct cros_ec_device *ec = dev_get_drvdata(ec_device);
+	if (ec->cmd_readmem) {
+		// Count the number of fans
+		size_t fan_count;
+		if (ec_count_fans(&fan_count) < 0) {
+			dev_err(dev, DRV_NAME ": failed to count fans.\n");
+			return -EINVAL;
+		}
+		// NULL terminates the list after the last detected fan
+		fw_hwmon_attrs[fan_count * FW_ATTRS_PER_FAN] = NULL;
+
+		data->hwmon_dev = hwmon_device_register_with_groups(
+			dev, DRV_NAME, NULL, fw_hwmon_groups);
+		if (IS_ERR(data->hwmon_dev))
+			return PTR_ERR(data->hwmon_dev);
+
+	} else {
+		dev_err(dev, DRV_NAME ": fan readings could not be enabled for this EC %s.\n",
+		FRAMEWORK_LAPTOP_EC_DEVICE_NAME);
+	}
+
 	battery_hook_register(&framework_laptop_battery_hook);
 
 	return ret;
 }
 
 static int framework_remove(struct platform_device *pdev)
 {
+	struct device *dev;
+	struct framework_data *data;
+
+	dev = &pdev->dev;
+
+	data = devm_kzalloc(dev, sizeof(*data), GFP_KERNEL);
```

Comment:
Doing this during `...remove` wil lactually allocate a _new_ empty `framework_data` (!) instead of getting the one associated with the current device.

I think the way to do this is to use `data = (framework_data*)platform_get_drvdata(pdev);` and check whether it is set. That will let you reuse the one you created in `framework_probe`.

Reply:
Ahh thanks! I was a little confused how that all worked, but that makes way more sense. I believe I've fixed it up now!